### PR TITLE
Adding "copy" as a parameter to klines. default to True

### PR DESCRIPTION
**Pull request by mishaker**
Will be useful for logs and slippage calculation between TA and real price at buy moment. more efficient than returning the whole DF.

### CODE REVIEW

File: freqtrade/tests/exchange/test_exchange.py [hunk 1]
```diff
@@ -814,6 +814,9 @@ def test_refresh_tickers(mocker, default_conf, caplog) -> None:
         assert isinstance(exchange.klines(pair), DataFrame)
         assert len(exchange.klines(pair)) > 0
 
+    # test last kline close price
+    assert exchange.last_kline_close('XRP/ETH') == 4
```
Comment (xmatthias) [inline freqtrade/tests/exchange/test_exchange.py]: i guess you'd want to test this with an invalid pair too ...

File: freqtrade/exchange/__init__.py [hunk 2]
```diff
@@ -164,6 +164,12 @@ def klines(self, pair: str) -> DataFrame:
         else:
             return None
 
+    def last_kline_close(self, pair: str):
```
Comment (xmatthias) [inline freqtrade/exchange/__init__.py]: you should also add the return type hint

**Comment by xmatthias** [comment]
thinking about how you're going to use it a bit more I'm not certain if adding a kwarg to `def klines(self, pair: str) -> DataFrame:` with `copy=True` would not be a better way.
you could (for specific cases, which don't modify the dataframe) return the original klines, not dataframe.

The copy is meant as a security when passing the dataframe to user-supplied code (strategy) - as we have no idea what happens there (someone could drop close for example  - in which case your code would fail) - but for internal "read only" operations the original DF would be perfectly fine.

**Comment by mishaker** [comment]
yep your approach is better ! aligning the PR with it ...

File: freqtrade/exchange/__init__.py [hunk 3]
```diff
@@ -158,9 +158,9 @@ def id(self) -> str:
         """exchange ccxt id"""
         return self._api.id
 
-    def klines(self, pair: str) -> DataFrame:
+    def klines(self, pair: str, copy=True) -> DataFrame:
         if pair in self._klines:
-            return self._klines[pair].copy()
```
Comment (xmatthias) [inline freqtrade/exchange/__init__.py]: would you mind adding a test to "confirm" this - i think you can do this easily by using 
```
assert id(exchange.klines('pair') != id(exchange.klines('pair')
assert id(exchange.klines('pair') != id(exchange.klines('pair', copy=True)
assert id(exchange.klines('pair', copy=True) != id(exchange.klines('pair', copy=True)
```
(note that the above is mock-code, so syntax-errors may have occured...

Comment (mishaker) [reply freqtrade/exchange/__init__.py]: "is" and ''is not" do the job I think ...

**Pull request merged**